sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -574,7 +574,8 @@ object ScalaReflection extends ScalaReflection {
               "cannot be used as field name\n" + walkedTypePath.mkString("\n"))
           }
 
-          val fieldValue = Invoke(inputObject, fieldName, dataTypeFor(fieldType))
+          val fieldValue = Invoke(inputObject, fieldName, dataTypeFor(fieldType),
+            returnNullable = inputObject.nullable || !fieldType.typeSymbol.asClass.isPrimitive)
```

Comment:
we use just write `returnNullable = !fieldType.typeSymbol.asClass.isPrimitive` and check `inputObject.nullable` in `Invoke.nullable`